### ISSUE TITLE
Aggregate by read ID for demultiplexing - UUID not found in database

### ISSUE THREAD

**Issue by SophieS9** [nanoporetech/megalodon#126]
Hi Marcus, following on from your suggestion above on how to deal with multiplexed data here https://github.com/nanoporetech/megalodon/issues/43, I've hit an issue with trying to run aggregation based on read IDs. I've run Megalodon on my data as follows:
```
megalodon --outputs basecalls mappings mod_basecalls mod_mappings mods --reference $ref --processes 8 --output-directory $out --devices cuda0 --guppy-server-path /opt/soft
ware/uoa/2021/apps/guppy/ont-guppy-gpu_4.4.2/bin/guppy_basecall_server --guppy-timeout 60.0 --overwrite $in
```
Followed by:
```
megalodon_extras per_read_text modified_bases $out
```
(I know that this could have been combined with the previous step but decided to run separate).
Then demultiplexed as follows:
```
guppy_barcoder -i $in --save_path $out --config configuration.cfg  --barcode_kits EXP-NBD114 --trim_barcodes
```
Got my read_IDs using the barcoding_summary.txt file:
```
#!/bin/bash

IDs=(
barcode14
barcode15
barcode16
barcode17
barcode18
barcode19
barcode20
barcode21
)

for i in "${IDs[@]}"; do
	cat barcoding_summary.txt | awk -v X=$i '($2==X){print $1}' > ${i}_readIDs.txt
done
```
And then finally trying to run aggregate by read_id:
```
megalodon_extras aggregate run --outputs mods --megalodon-directory $mega_out --output-suffix $barcode --read-ids-filename $read_ids
```
And I'm getting the following error:
```
Traceback (most recent call last):
  File "/uoa/home/s10ss5/.local/lib/python3.7/site-packages/megalodon/mods.py", line 448, in get_read_dbid
    read_dbid = self.uuid_to_dbid[uuid]
KeyError: 'af201f30-fa9c-48fc-806b-451835be54a7'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/software/uoa/2019/conda/anaconda3/envs/megalodon-2.2.10/bin/megalodon_extras", line 11, in <module>
    sys.exit(_main())
  File "/uoa/home/s10ss5/.local/lib/python3.7/site-packages/megalodon_extras/__main__.py", line 165, in _main
    module._main(args)
  File "/uoa/home/s10ss5/.local/lib/python3.7/site-packages/megalodon_extras/aggregate_run.py", line 36, in _main
    args.aggregate_batch_size)
  File "/uoa/home/s10ss5/.local/lib/python3.7/site-packages/megalodon/aggregate.py", line 327, in aggregate_stats
    valid_read_dbids.add(mods_db.get_read_dbid(read_id))
  File "/uoa/home/s10ss5/.local/lib/python3.7/site-packages/megalodon/mods.py", line 455, in get_read_dbid
    'UUID not found in database: {}'.format(uuid))
megalodon.megalodon_helper.MegaError: UUID not found in database: af201f30-fa9c-48fc-806b-451835be54a7
```
I've checked simply using grep and that read ID is present in my basecalled fastq file but doesn't appear in my "per_read_modified_base_calls.txt" file so I'm assuming that even though it was basecalled it's not modified? Is there a way to skip this read ID in the aggregation? Or would you recommend a filtering step here removing all read IDs that aren't in the per_read_modified_base_calls.txt before aggregating? 

EDIT - I've searched for this read ID in the log.txt and it was a read that failed to align to the reference genome.

Thanks! 

Sophie

**Comment by marcus1487** [comment]
Megalodon should likely warn in this setting instead of exiting with an error. I'll have a look at a proper fix for this.

**Comment by SophieS9** [comment]
Thanks! I'm putting together something simple using the log.txt file to remove any read IDs which have failed before running the aggregate, but having megalodon just produce a warning would save an unneeded step!

**Comment by marcus1487** [comment]
Fixed in 2.3.0 release.

**Comment by ginolhac** [comment]
Of note, this is not fixed in version 2.3.1 nor 2.3.3. I had to remove by hand read-ids that were not aligned

**Comment by marcus1487** [comment]
Thank you very much for the bug notification. It looks like this behavior got reverted (or was never really fixed). I've just pushed a fix to github that should resolve this issue. This will be included in the next release, but to use this fix now you can re-install megalodon from github via `pip install git+https://github.com/nanoporetech/megalodon`
